### DISCUSSION TITLE
Context propagation using messages (message broker)

### DISCUSSION THREAD

**Discussion by alexarieta** [open-telemetry/opentelemetry-cpp#780] [Q&A]
Hello

In my project we are looking into integrating opentelemetry for distributed tracing, where we have several nodes (PCs) on one side running in c# and on the other c++. We use a message broker to communicate between the applications, so there is no direct http connection between them.

Do you have an example on how to propagate the context through messages payload?

Thank you

**Comment by lalitb** [comment]
There is no example as of now, but you can custom write your own Propagator by implementing [`Text Map Propagator`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#textmap-propagator) for your message broker. This propagator should provide the functionality to serialize and deserialize SpanContext of currently running Span so that it can be sent/received through the message broker. In C++, you can do this by inhering from [TextMapPropagator ](https://opentelemetry-cpp.readthedocs.io/en/latest/otel_api/classopentelemetry_1_1context_1_1propagation_1_1TextMapPropagator.html#exhale-class-classopentelemetry-1-1context-1-1propagation-1-1textmappropagator) class and implementing Inject/Extract methods. And the similar thing needs to be done for C#.

**Reply by alexarieta** [reply]
Thank you
